src/backend/crossterm.rs

```diff
@@ -127,6 +119,194 @@ where
+    /// # std::io::Result::Ok(())
+    /// ```
+    pub fn into_terminal_with_defaults(self) -> io::Result<Terminal<Self>> {
+        let backend = self.with_raw_mode()?.with_alternate_screen()?;
+        Terminal::new(backend)
+    }
+
+    /// Enables raw mode for the terminal.
+    ///
+    /// Returns an [`io::Result`] containing self so that it can be chained with other methods.
+    ///
+    /// Raw mode is restored when the `CrosstermBackend` is dropped.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use ratatui::backend::CrosstermBackend;
+    ///
+    /// let backend = CrosstermBackend::stdout().with_raw_mode()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    pub fn with_raw_mode(mut self) -> io::Result<Self> {
+        self.restore_raw_mode_on_drop = true;
+        terminal::enable_raw_mode()?;
+        Ok(self)
+    }
+
+    /// Enables raw mode for the terminal and switches to the alternate screen.
+    ///
+    /// Returns an [`io::Result`] containing self so that it can be chained with other methods.
+    ///
+    /// Alternate screen is restored when the `CrosstermBackend` is dropped.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use ratatui::backend::CrosstermBackend;
+    ///
+    /// let backend = CrosstermBackend::stdout().with_alternate_screen()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    pub fn with_alternate_screen(mut self) -> io::Result<Self> {
+        self.restore_alternate_screen_on_drop = true;
+        execute!(self.writer, terminal::EnterAlternateScreen)?;
+        Ok(self)
+    }
+
+    /// Enables mouse support for the terminal.
+    ///
+    /// Returns an [`io::Result`] containing self so that it can be chained with other methods.
+    ///
+    /// Mouse support is restored when the `CrosstermBackend` is dropped.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use ratatui::backend::CrosstermBackend;
+    ///
+    /// let backend = CrosstermBackend::stdout().with_mouse_support()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    pub fn with_mouse_support(mut self) -> io::Result<Self> {
+        self.restore_mouse_capture_on_drop = true;
+        execute!(self.writer, crossterm::event::EnableMouseCapture)?;
+        Ok(self)
+    }
+
+    /// Resets the terminal to its default state.
+    ///
+    /// Disables raw mode, disables mouse capture, and leaves the alternate screen.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use std::io::stdout;
+    ///
+    /// use ratatui::backend::CrosstermBackend;
+    ///
+    /// CrosstermBackend::reset(stdout())?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    pub fn reset(mut writer: W) -> io::Result<()> {
+        terminal::disable_raw_mode()?;
+        execute!(
+            writer,
+            terminal::LeaveAlternateScreen,
+            crossterm::event::DisableMouseCapture
+        )
+    }
+}
+
+impl<W: Write> Drop for CrosstermBackend<W> {
```

Comment:
Drop implementations are kinda annoying to work with. It also creates another duplicate way of doing things. Reset or drop?
I would go for the explicit reset. It's also quite obvious when the reset wasn't done when stopping the process. Having a reset and for some reason accidentally using the drop implementation on error cases for example might result in different behavior which might also be confusing.

Reply:
One possible approach to this might be to add reset to the Backend trait with no parameter and have a version with the parameter on the method which will be called on CrosstermBackend. This overloads the name, but it's reasonable-ish.

Reply:
Then reset should probably be removed? When users want to do it explicitly they can use `drop(backend)` then.

Reply:
That's literally not possible within a panic handler which is where this is necessary.

Reply:
Comment added in cd525c335

Reply:
So basically, reset is never used except for panic handlers? The single way of cleaning up is the drop implementation.

Reply:
Yes. Reset should be used in situations where you don't have access to the backend instance. Panic handlers is one such example. I can conceive of some others, but the panic handler is the absolutely necessary part of this.